### DISCUSSION TITLE
Guidance / support for client side auth

### DISCUSSION THREAD

**Discussion by cezapata** [community/community#7227] [Codespaces]
I would like to know if there is guidance to set up an OAuth workflow alongside Codespaces.

A use case I can think of is having a Single-page application that uses a client-side auth library with the [implicit grant flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-implicit-grant-flow) to authenticate.

As part of the contract, we are required to provide a reply URL which needs to map with the running codepsace URL. This would be a challenge since Codespace URLs are dynamic.

I think that running this on VS Code Desktop is not an issue, since we can use localhost or our hosts file to map to a known URL, but running the app in the browser would be difficult.

**Comment by joshspicer** [comment]
(we chatted about this already)

When running a codespace in VSCode the ports you forward are available through `localhost:PORT`, but I agree it would be handy to always have a deterministic port forward URI regardless which client you're using (web comes to mind, also probably SSH?)

**Reply by Chuxel** [reply]
There is aa deterministic way to do this if you allow https://*.githubpreview.dev to be a valid URI in your test authentication flow. Even when using VS Code, forwarded ports are available using the web URI.

You can programmatically determine the URL thanks to a `CODESPACE_NAME` environment variable that is in every codespace. 

Port forwarded urls are always:

```
https://${CODESPACE_NAME}-<port>.githubpreview.dev
```

Where <port> is the port number you expect to receive the authentication flow.  You should be able to use this consistently to get a callback URL even when using VS Code desktop.

The `CODESPACE_NAME` variable would not be set if you're outside of a codespace and nee to vary your logic for some reason. There's also a `CODESPACES=true` variable for similar purposes.

**Reply by cezapata** [reply]
Thanks @Chuxel. This would be feasible from a technical perspective. I am wondering, however, whether this could expose an auth provider to a vulnerability by allowing any codespaced app as a reply URL (I am not a security expert, and just wondered whether it could increase our attack surface area).

If this was an actual security issue, I think a mitigation strategy would be to allow subdomains in the URL that can map to something like a team or organization name, like:

```
https://${CODESPACE_NAME}-<port>.<organization>.githubpreview.dev
```

This could allow us to uniquely identify the apps through our auth service provider using wildcards.

**Reply by Chuxel** [reply]
It would, but its conceptually the same risk as allowing `localhost`. Anyone could do this from their local machine when localhost is allowed - this is Codespace's localhost. I would definitely not use a production setup in either of these cases but instead have a test one (specifics here varying a bit depending on your auth implementation).

That said, an org specific URI is an interesting feature request.

**Reply by cezapata** [reply]
I hadn't thought about it that way but that is true 😄. Thanks for the suggestion!
